LowerToBackendContract: Explicitly error out on unimplemented operator

mgehre-amd opened the pull request:
As discussed on discord [0], this PR will emit more direct errors like `unsupported by backend contract: Unimplemented operator 'aten.clamp.Tensor'` instead of diagnosing unknown ranks or missing value conversion.


[0] https://discordapp.com/channels/636084430946959380/742573221882364009/1083826291301679225

vivekkhandelwal1 committed:
[MLIR][TORCH] Add canonicalizer for aten.new_empty_strided op

Signed-Off By: Vivek Khandelwal<vivek@nod-labs.com>

GlebKazantaev committed:
Fix OptionalCType class name (#1779)

* Fix OptionalCType class name

* Rmove LTC xfail tests

ashay committed:
CI: miscellaneous fixes for Release builds (#1781)

- Use v3 of actions/checkout, since the version we use (v2) uses
   Node.js 12, which is deprecated by GitHub.

 - Source the PowerShell venv sctipt (instead of the bash sript) since
   the calling script is a PowerShell script.  Without this, the build
   doesn't use venv at all.

 - Make the build dependencies in whl-requirements.txt (used by
   setup.py) match those in requirements.txt.  To that end, this patch
   creates a build-requirements.txt that is referenced by
   requirements.txt and whl-requirements.txt.

vivekkhandelwal1 committed:
build: manually update PyTorch version
Set PyTorch and TorchVision version to nightly release 2023-01-08.

Signed-Off By: Vivek Khandelwal<vivek@nod-labs.com>

GlebKazantaev committed:
Added support for aten::norm.ScalarOpt_dim (#1774)

* Added support for aten::norm.ScalarOpt_dim

* Disable NormalizeModule_basic for linalg

navahgar committed:
Fix unsqueeze in Torch to Tosa conversion (#1780)

ashay committed:
build: update llvm tag to de3f0f7f (#1789)

Credit to @vivekkhandelwal1 for finding the necessary changes.

Summary of changes:

 - Switch Tosa_IntArrayAttr[N], Tosa_IntArrayAttrUpto[N] to DenseI64ArrayAttr.

 - Replace kNoIterationLimit with kNoLimit. (https://reviews.llvm.org/D140525)

 - Add dependency on MhloPasses when MHLO is enabled

 - Specify result type when using mhlo::DotOp

ashay committed:
[TOSA] Add LeakyReLU conversion pass (#1790)

* feat(TorchToTOSA): LeakyReLU legalization

* test(LeakyReLU): Add LIT test and enable e2e test

Co-authored-by: Philipp Braun <philipp.braun@amd.com>

vivekkhandelwal1 committed:
[MLIR][TORCH] Add decomposition for prims.var and prims.sqrt op

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

Commit:
update PyTorch version to 2.0.0.dev20230111

asaadaldien committed:
Use global for MLIRTorchOpsIncGenTdFiles (#1784)

li-plus committed:
[LINALG][TOSA][MHLO] Add e2e support for aten bitwise ops (#1753)

Commit:
update PyTorch version to 2.0.0.dev20230112

ramiro050 committed:
Fix sign-compare warnings in TosaLegalizeCommon.cpp (#1794)

Commit:
update PyTorch version to 2.0.0.dev20230113

Commit:
update PyTorch version to 2.0.0.dev20230114

vivekkhandelwal1 committed:
build: manually update PyTorch version

Set PyTorch and TorchVision version to nightly release 2023-01-15.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

vivekkhandelwal1 committed:
[torchdynamo] Add native_group_norm and split op to the decomp list

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

vivekkhandelwal1 committed:
[MLIR][TORCH] Add decomposition for aten.randn_like op

This commit decomposes aten.randn_like op into aten.randn.generator op.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

Commit:
update PyTorch version to 2.0.0.dev20230116

makslevental committed:
Expose metadata of torch-mlir types (plus verify DictType key and value types). (#1785)

Commit:
update PyTorch version to 2.0.0.dev20230117

li-plus committed:
Fix empty tensor when select -1 (#1787)

qingyunqu committed:
[MHLO] simplify aten.frobenius_norm.dim's lowering (#1800)

Commit:
update PyTorch version to 2.0.0.dev20230118

Commit:
update PyTorch version to 2.0.0.dev20230119

Commit:
[TOSA] Add aten.Index.Tensor support (#1771)

Commit:
update PyTorch version to 2.0.0.dev20230120

ramiro050 committed:
Make `getTypeForScalarType` safer by returning `FailureOr<Type>` (#1814)

One of the potential values for a `torch_upstream::ScalarType` is
`Undefined`. This means that conversion of a `ScalarType` to another
type is a computation that can fail. To enforce handling of the
failure case, this commit makes the two helper functions that convert
`ScalarType`s into other types return `failure()` when the
`ScalarType` is `Undefined`.

Commit:
[TOSA] Add aten.view shape -1 support (#1815)

Commit:
update PyTorch version to 2.0.0.dev20230121

vguerra committed:
Using `ArrayRef` deduction guides. (#1809)

`llvm::makeArrayRef` is now deprecated and can be
replaced by the newly introduced `ArrayRef` deduction guides.

Fixes: #1808

Co-authored-by: Victor Guerra <vm.guerramoran@criteo.com>

powderluv committed:
Use delvewheel to package windows DLLs (#1820)

vivekkhandelwal1 committed:
build: manually update PyTorch version

Set PyTorch and TorchVision version to nightly release 2023-01-22.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

vivekkhandelwal1 committed:
[torchdynamo] Add default decomposition for ops in the dynamo backend

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

Commit:
update PyTorch version to 2.0.0.dev20230123

ashay committed:
CI: build torch-mlir python package for Python v3.8 (#1827)

Previously, torchvision had not released WHL files for Python v3.8,
causing failures in torch-mlir python package builds, so we had disabled
building for Python v3.8.

Now that the WHL files are back, this patch re-enables v3.8 builds.

eric-k256 committed:
update llvm to d23516e9ad477527a9db4d06b1fa9566680ac67c (#1812)

Rename BlockAndValueMapping to IRMapping
Moved PrimTupleConstructOp type validation to its own verifier as the
tablegen version does not work for a combination of variadic input and
non-variadic output.

GlebKazantaev committed:
Fix JIT schema matching for when ListType is used (#1826)

Commit:
update PyTorch version to 2.0.0.dev20230124

ramiro050 committed:
build: update llvm tag to 9acc2f37 (#1828)

This commit makes the following changes:

- Update dialects to use fold API `kEmitFoldAdaptorFolder` and update
signature of `fold` methods (see PSA
https://discourse.llvm.org/t/psa-new-improved-fold-method-signature-has-landed-please-update-your-downstream-projects/67618)
- Replace `makeArrayRef` with `ArrayRef` (see
https://reviews.llvm.org/D140896)
- Remove `TypeRange{}` arg from `b.create<scf::IfOp>` since builder no
longer takes that argument
- Make `func`s in `Torch/invalid.mlir` private, since symbol
declarations cannot be public. (see https://discourse.llvm.org/t/rfc-symbol-definition-declaration-x-visibility-checks/2140)

GlebKazantaev committed:
Enable VerifyBackendContract in LTC backend (#1798)

* Enable VerifyBackendContract in LTC backend

* Update VerifyBackendContract pass

* Move convert_scalar_implicit to jit_utils

* Rename VerifyBackendContract to VerifyBackendContractNoDecompositions

* Update verify-backend-contract-error.mlir test

mgehre-amd committed:
[TorchToLinalg] Lower AtenRoundOp to math::RoundEvenOp (Fixes #1811) (#1823)

[TorchToLinalg] Lower AtenRoundOp to math::RoundEvenOp (Fixes #1811)

Commit:
update PyTorch version to 2.0.0.dev20230125

wangkuiyi committed:
Update README.md and put Python 3.10 a prerequisite (#1821)

* Update README.md and put Python 3.10 a prerequisite

* Update README.md

* Follow up with comments

* Update README.md

Commit:
update PyTorch version to 2.0.0.dev20230126

Commit:
update PyTorch version to 2.0.0.dev20230127

Commit:
[TOSA] Add to.dtype i1 to i64 (#1830)

Commit:
update PyTorch version to 2.0.0.dev20230128

Commit:
update PyTorch version to 2.0.0.dev20230129

li-plus committed:
[MHLO] Support aten.cumsum op in mhlo backend (#1825)

Commit:
update PyTorch version to 2.0.0.dev20230130

silvasean committed:
Update LLVM and MHLO submodules.

Week of 01/30/2023:

Green LLVM commit: e31ee6417c33a6e2f0e8440b1a86d5365279ad68
Green MHLO commit: c2a6f4064d426567b9ef7b0d29d5ab86dc7b2b02 (branch greencommit/2023-01-30-e31ee641)

Commit:
update PyTorch version to 2.0.0.dev20230131

vivekkhandelwal1 committed:
[MLIR][TORCH] Add support for clone op with channels last memory format

Fixes https://github.com/llvm/torch-mlir/issues/1829

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

Commit:
update PyTorch version to 2.0.0.dev20230201

ashay committed:
Revert "update PyTorch version to 2.0.0.dev20230201" (#1842)

This reverts commit ca400e02cecb8d7efc062fa895a7fcdd8917b84e, since the
(upstream) Windows build of torchvision has an incorrect dependency.

ashay committed:
mhlo: migrate conversion to stablehlo (#1840)

This patch replaces all MHLO operations with their StableHLO
counterparts and adds a validation pass to ensure that no MHLO operations
remain before translating all Stablehlo operations to the MHLO dialect
for further lowering to the Linalg dialect.

This patch also updates all lit tests so that they refer to the
`convert-torch-to-stablehlo` pass and so that they check for StableHLO
operations.

Commit:
update PyTorch version to 2.0.0.dev20230202

li-plus committed:
Add aten.bucketize op and its decomposition (#1834)

cetiniz committed:
Add in-place variant of torch.scatter_add (#1836)

vivekkhandelwal1 committed:
build: manually update PyTorch version

Set PyTorch and TorchVision version to nightly release 2023-02-05.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

vivekkhandelwal1 committed:
[MLIR][TORCH] Add support for integer type input for max.dim op

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

qingyunqu committed:
[MHLO] move AtenTanhOp to ConvertAtenUnaryFPOnlyPatten and add sin/cos/ceil/floor pattern (#1847)

vivekkhandelwal1 committed:
build: manually update PyTorch version

Set PyTorch and TorchVision version to nightly release 2023-02-06.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

Commit:
build: update llvm tag to ba8b8a73f (#1856)

* build: update llvm tag to ba8b8a73f

* skip ConvolutionModule2DGroups_basic

cetiniz committed:
Add aten.scatter_reduce op definition (#1846)

vivekkhandelwal1 committed:
build: manually update PyTorch version

Set PyTorch and TorchVision version to nightly release 2023-02-07.
Fixes https://github.com/llvm/torch-mlir/issues/1792.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

Commit:
update PyTorch version to 2.0.0.dev20230208

Commit:
update PyTorch version to 2.0.0.dev20230209

Commit:
[TOSA] Fix broadcast_to input and output different shape support (#1855)

Groverkss committed:
Add compile_commands.json to gitignore (#1867)

Commit:
update PyTorch version to 2.0.0.dev20230210

powderluv committed:
Python 3.11 support (#1848)

* Python 3.11 support

* test without torchvision

* Update pytorch-requirements.txt

* Update buildRelease.yml

* Update action.yml

* Update install_macos_deps.sh

* Update build_macos_packages.sh

qingyunqu committed:
[Torch Dialect] add folder for prim.min.int (#1864)

ZihengJiang committed:
[MHLO] Support AtenMaskedFillScalar (#1839)

* [MHLO] Support MaskedFillScalar.

* Update.

* Update.

* Update.

---------

Co-authored-by: Ziheng Jiang <ziheng.jiang@bytedance.com>

qingyunqu committed:
[Torch Dialect] add folder for aten.Int.float (#1863)

Commit:
update PyTorch version to 2.0.0.dev20230211

Commit:
update PyTorch version to 2.0.0.dev20230212

Commit:
update PyTorch version to 2.0.0.dev20230213

makslevental committed:
WIP: No PyTorch dep (#1854)

powderluv committed:
test torch / torchvision compat (#1877)

ashay committed:
python: separate build- and test-related pip dependencies (#1874)

We want to ensure that pip packages required for building torch-mlir
should be included in the dependencies of torch-mlir, but we don't want
the pip packages required for _testing_ of torch-mlir to be included
among the dependencies.  To be able to specify and install one set of
dependencies and not the other, this patch separates the pip packages
into two files: build-requirements.txt and test-requirements.txt.

This patch also updates references to the requirements.txt file so that
CI builds that run end-to-end tests install test-related pip
dependencies while everything else (including WHL builds) sticks to just
the build-related pip dependencies.

Despite this change, this patch should not affect a torch-mlir
developer's workflow.  More precisely, since this patch makes the
top-level requirements.txt file refer to both build-requirements.txt and
test-requirements.txt files, a torch-mlir developer should be able to
continue referring to the requirements.txt file without any impact.

Commit:
[TOSA] aten.index.tensor multiple indexes support (#1868)

vivekkhandelwal1 committed:
[MLIR][TORCH] Add decomposition for aten.bernoulli.p op (#1882)

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

powderluv committed:
Update buildAndTest.yml (#1881)

* Update buildAndTest.yml

* Update oneshotSnapshotPackage.yml

* Update buildRelease.yml

* Update RollPyTorch.yml

* Update oneshotSnapshotPackage.yml

* Update buildAndTest.yml

Commit:
[TOSA] Add todtype f64 and i1 to f32 support (#1845)

ZihengJiang committed:
[StableHLO] Add support for AtenPowTensorScalar. (#1883)

* [MHLO] Add support for AtenPowTensorScalar.

* Update.

---------

Co-authored-by: Ziheng Jiang <ziheng.jiang@bytedance.com>

ashay committed:
CI: install `unzip` before using it (#1893)

The RollPyTorch action needs the `unzip` command to peek into WHL files
for fetching metadata.  This patch makes sure that the command is
installed before referencing it.

vivekkhandelwal1 committed:
build: manually update PyTorch version

Set PyTorch and TorchVision version to nightly release 2023-02-20.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

qingyunqu committed:
[Torch] decompose aten.norm.ScalarOpt_dim to aten.linalg_vector_norm (#1849)

ramiro050 committed:
Replace `torch.rand` and `torch.randn` in e2e tests with `tu.rand` (#1890)

Random tensors used in e2e tests should be created using the
`TestUtils` object passed to the registered test case to ensure that
the compiled module and the golden trace receive the same tensors as
input. This commit changes all the cases of `torch.rand` and
`torch.randn` to use the `TestUtils` instead.

cetiniz committed:
[Torch Dialect] Scatter reduce lowering (#1884)

- Lowers the torch.scatter_reduce to linalg_on_tensors dialect.
- Includes support for "sum", "prod", "amax", "amin" and "mean".

Commit:
build: update llvm tag to 5e111eb275

- Update llvm tag to 5e111eb275eee3bec1123b4b85606328017e5ee5
- mhlo now points to a99159c45ee5c497f8dce01eff807a6d57629b61

Signed-Off-by: Gaurav Shukla <gaurav@nod-labs.com>

prateekgu-cerebras committed:
[TORCH-MLIR] Add ODS for aten.clamp.Tensor op. (#1894)

This commit adds the ODS definition for the aten.clamp.Tensor op.

Signed-off-by: Prateek Gupta <prateek.gupta2@cerebras.net>

ashay committed:
CI: clear workspace directory before checkout (#1900)

We have recently started seeing errors like:

```
  Synchronizing submodule url for 'externals/llvm-project'
  Synchronizing submodule url for 'externals/mlir-hlo'
  /usr/bin/git -c protocol.version=2 submodule update --init --force --depth=1
  Error: fatal: Unable to create '/home/anush/actions-runner/_work/torch-mlir/torch-mlir/.git/modules/externals/llvm-project/index.lock': File exists.
```

As a workaround, this patch removes the workspace directory before the
checkout step.

erl4ng committed:
Update README.md (#1880)

* Update README.md

refined formal definition

* Update README.md

toned down "revolutionary" to novel.

asaadaldien committed:
[Bazel] Fix clang version, missing mhlo deps and add reqs files (#1898)

* Fix clang version, missing mhlo deps and add reqs files

ashay committed:
CI: use `sudo` to remove Docker-created files (#1905)

vivekkhandelwal1 committed:
build: manually update PyTorch version

Set PyTorch and TorchVision version to nightly release 2023-02-27.
This commit also adds the lowering for aten.add and aten.Float.Scalar op.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

vivekkhandelwal1 committed:
build: update llvm tag to b5bf6f63

- Remove `useFoldAPI`.
- Update tags
  llvm: b5bf6f6392a3408be1b7b7e036eb69358c5a2c29
  mhlo: bb1ed5b47c036993b954e881cce7c7aab00f9962

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

qingyunqu committed:
[Torch Dialect] add aten.pow.int_float op and it's folder (#1872)

torch-mlir committed:
update PyTorch version to 2.0.0.dev20230301

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230302

qingyunqu committed:
[Torch Dialect] add folder for aten.sub.float (#1871)

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230303

ramiro050 committed:
Handle uninitialized lattice elements in RefineTypes (#1911)

The data-flow analysis does not always propagate information to the
entire graph. This results in some lattice elements being
uninitialized. Currently the lattice elements are not checked to see
if they are uninitialized before rewriting the graph, potentially
resulting in invalid IR (see
https://github.com/llvm/torch-mlir/issues/1896).

This commit adds handling for uninitialized lattice elements.

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230304

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230305

makslevental committed:
Add `torch-mlir-no-jit-importer` build case for mac os wheels (#1902)

* add flags to setup.py for out-of-tree build

* - fix build_ext bug
- add wheels script cases for mac wheels

makslevental committed:
add jit,no-jit release matrix (#1916)

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230306

ramiro050 committed:
Fix handling of non-int tensors in `getScalarValue` (#1914)

The current implementation of `getScalarValue` does not check that the
input to a `ValueTensorLiteralOp` is an i64 before extracting the
value, and it does not check that the result type of the
`PrimNumToTensorScalarOp` is also an i64. This leads to crashes or
invalid IR generated when the `input` is something other than an i64
tensor or `!torch.int`.

This commit addresses those issues. In addition, the function
`getScalarValue` is renamed to `getScalarIntValue` to make it clear
that it *only* extracts scalar integers.

PriyaBSavithiri committed:
Add HardtanhBackward TOSA and LINALG support (#1721)

zhekunz2 committed:
[Torch Dialect] add RSub, ScalarImplicit canonicalize (#1899)

* add rsub, scalarimplit canonicalizer

* reformat

* address comments

* fix bug

* fix test

* Update elementwise.py

* resolve merge conflict

* change to 3

* change to 3

* real fix

* fix name

* add torchdynamo fail test

---------

Co-authored-by: zhekun.zhang <zhekun.zhang@bytedance.com>

makslevental committed:
- rename no-jit -> core (#1920)

- add windows release

eric-k256 committed:
Update LLVM tag to 21f4b84c (#1918)

Update PassManager C interface to use mlirPassManagerRunOnOp
Update python calls to PassManager to also use operation instead of
module

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230307

chenkanhw committed:
Adding stablehlo dialects support for torch-mlir-opt tool (#1887)

* Adding stablehlo dialects support for torch-mlir-opt tool.

* Update torch-mlir-opt.cpp

Fixed the build error according to build configuration for macOS.

chenkanhw committed:
Fix the error of type casting in dynamo example (#1860)

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230308

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230309

ZihengJiang committed:
[TORCH] Improve type refinement for aten.cat. (#1908)

* [TORCH] Fix type refinement for aten.cat.

* Add test.

* Address comments.

* Update.

* Update.

* Update.

* Update.

* Update.

---------

Co-authored-by: Ziheng Jiang <ziheng.jiang@bytedance.com>

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230310

ramiro050 committed:
Fix deprecation warnings for `isOneValue` and `getAllOnesValue` (#1928)

The functions `isOneValue` and `getAllOnesValues` are
deprecated. `isOne` and `getAllOnes` should be used instead.

gpetters94 committed:
Add a new RecomposeComplexOps pass, fold slice+copy_ into indeX_put_ (#1901)

ramiro050 committed:
Expand definition of tensor subtype to include shape/dtype info (#1929)

Currently, the op `torch.tensor_static_info_cast` will not get
canonicalized away if the result type has any shape or dtype
information. This is because `isValidSubtype` only returns true when
the tensor types being compared are exactly the same or the supertype
has no shape and dtype information. Being unable to canonicalize away
the `torch.tensor_static_info_cast` gets in the way of further
optimizations, such as shape propagation.

This commit improves `isValidSubtype` by adding logic that compares
the shapes and dtypes of the two tensor types to determine of one type
is indeed a valid subtype of the other.

Fixes https://github.com/llvm/torch-mlir/issues/1926

li-plus committed:
Support aten.stack op and decompose it into unsqueeze & cat (#1747)

qingyunqu committed:
[e2e] fix stack e2e test typo (#1931)

sjain-stanford committed:
[Bazel] Add StableHLO registration dependency (#1937)

Fixes a broken bazel build from https://github.com/llvm/torch-mlir/pull/1887 causing the following build error:

```
ERROR: /root/.cache/bazel/_bazel_root/b89349c08f7224396763d14fe35cba11/external/torch-mlir/BUILD.bazel:819:10: Compiling tools/torch-mlir-opt/torch-mlir-opt.cpp failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-16/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer ... (remaining 366 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
external/torch-mlir/tools/torch-mlir-opt/torch-mlir-opt.cpp:16:10: fatal error: 'stablehlo/dialect/Register.h' file not found
#include "stablehlo/dialect/Register.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Target @torch-mlir//:torch-mlir-opt failed to build
```

GHA workflow with the fix:
https://github.com/sjain-stanford/torch-mlir/actions/runs/4421486154

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230315

lisaliu1 committed:
Constant pad nd to tosa (#1933)

* implemented lowering torch.aten.constant_pad_nd to tosa
* add constant_pad_nd e2e tests to TOSA_PASS_SET
* add PadModule_basic & PadWithNoneValModule_basic to TOSA_PASS_SET
---------

Co-authored-by: Lisa Liu <lingl@xilinx.com>

mgehre-amd committed:
LowerToBackendContract: Explicitly error out on unimplemented operator

ljfitz commented on test/Dialect/Torch/verify-backend-contract-unimplemented-op.mlir:
```diff
@@ -0,0 +1,10 @@
+  %0 = torch.tensor_static_info_cast %arg0 : !torch.vtensor<[3,5],f32> to !torch.vtensor<*,f32>
+  %1 = torch.copy.to_tensor %0 : !torch.tensor<*,f32>
+  // expected-error @+1 {{unsupported by backend contract: Unimplemented operator 'aten.clamp.Tensor'}}
+  %2 = torch.operator "aten.clamp.Tensor"(%1, %1, %none) : (!torch.tensor<*,f32>, !torch.tensor<*,f32>, !torch.none) -> !torch.tensor
```
Is `aten.clamp.Tensor` a real operation that might be implemented one day? If so, would it be legal to use a made up op name e.g. `an.unimplemented.op`?

mgehre-amd replied:
Good idea, I will adapt

ljfitz reviewed (approved):
Looks like a very nice change.

torch-mlir committed:
update PyTorch version to 2.1.0.dev20230316

ramiro050 committed:
RefineTypes aten.sum: check dtype exists before accessing methods (#1944)

This commit adds a check that `defaultDtype` exists in the RefineTypes
handling of `AtenSumOp` before accessing the method `isInteger`, which
crashes the program is `defaultDtype` is null.

The handling of `defaultDtype` is the same as the one used for the
`AtenSumDimIntListOp`.
